src/msg/Message.h

```diff
@@ -521,55 +520,21 @@ extern void encode_message(Message *m, uint64_t features, ceph::buffer::list& bl
 extern Message *decode_message(CephContext *cct, int crcflags,
                                ceph::buffer::list::const_iterator& bl);
 
-template <class MessageType>
-class MessageFactory {
-public:
-template<typename... Args>
-  static typename MessageType::ref build(Args&&... args) {
-    return typename MessageType::ref(new MessageType(std::forward<Args>(args)...), false);
-  }
-};
-
-template<class T, class M = Message>
-class MessageSubType : public M {
-public:
-  typedef boost::intrusive_ptr<T> ref;
-  typedef boost::intrusive_ptr<T const> const_ref;
-
-  static auto msgref_cast(typename M::ref const& m) {
-    return boost::static_pointer_cast<typename T::const_ref::element_type, typename std::remove_reference<decltype(m)>::type::element_type>(m);
-  }
-  static auto msgref_cast(typename M::const_ref const& m) {
-    return boost::static_pointer_cast<typename T::ref::element_type, typename std::remove_reference<decltype(m)>::type::element_type>(m);
-  }
-
-protected:
-template<typename... Args>
-  MessageSubType(Args&&... args) : M(std::forward<Args>(args)...) {}
-  virtual ~MessageSubType() override {}
-};
-
-
-template<class T, class M = Message>
-class MessageInstance : public MessageSubType<T, M> {
-public:
-  using factory = MessageFactory<T>;
-
-  template<typename... Args>
-  static auto create(Args&&... args) {
-    return MessageFactory<T>::build(std::forward<Args>(args)...);
-  }
-  static auto msgref_cast(typename Message::ref const& m) {
-    return boost::static_pointer_cast<typename T::ref::element_type, typename std::remove_reference<decltype(m)>::type::element_type>(m);
-  }
-  static auto msgref_cast(typename Message::const_ref const& m) {
-    return boost::static_pointer_cast<typename T::const_ref::element_type, typename std::remove_reference<decltype(m)>::type::element_type>(m);
-  }
-
-protected:
-template<typename... Args>
-  MessageInstance(Args&&... args) : MessageSubType<T,M>(std::forward<Args>(args)...) {}
-  virtual ~MessageInstance() override {}
-};
+namespace ceph {
+template<typename T> using ref_t = boost::intrusive_ptr<T>;
+template<typename T> using cref_t = boost::intrusive_ptr<T const>;
+template<class T, class U>
+boost::intrusive_ptr<T> ref_cast(const boost::intrusive_ptr<U>& r) noexcept {
```

Comment:
We can't return an lvalue? This creates a new reference which is suboptimal.

Reply:
i added an overload of this cast helper. it accepts `boost::intrusive_ptr<U>&&`. and changed some of the `ms_dispatch2()` to use it. but please note, not all callers can use this variant.

Reply:
I'm confused why we can't do something like:

https://github.com/ceph/ceph/pull/27325/commits/c507584feb85fb0c2bf3daad66130259458bcbce#diff-4f0a7af2bc469093c9fa5879e0c43100R101

your commit 733b2873c07660fb5cf76e6b9b3ab29cb8a602c9 mentions avoiding an "extra conversion" but I don't see what conversion you're referring to or why it's a bad thing.

Reply:
@batrick `boost::static_pointer_cast` is implemented using `static_cast`. the extra conversion *is* `boost::static_pointer_cast`. i guess it's just compile-time overhead, but the change is more symmetric this way.

Reply:
I didn't pipe up, but that was my question, too.

Matt

Reply:
@mattbenjamin does my comment answer your question?

Reply:
I see now that my version was actually returning a temporary and not an lvalue; doh! Maybe it's not even possible to convert `const boost::intrusive_ptr<Foo>&` to `const boost::intrusive_ptr<DerivedFoo>&` without creating a temporary ref?

Reply:
@batrick i assume that, by "a temporary ref", you meant we are

1. creating a different intrusive_ptr which increases the refcount of the raw pointer first, and 
1. and pass it to a function which expect this different intrusive_ptr.
1. after using it, we drop it on the floor. 

hence we are wasting CPU cycles. 

i agree with you. no, it's not possible, that's what the `const` is for. it forbids us to steal the ownership from `const boost::intrusive_ptr<DerivedFoo>&` for constructing a different smart pointer.

Reply:
@tchaikov I think so, yes;  thanks